sdk/include/opentelemetry/sdk/metrics/controller.h

```diff
@@ -99,16 +97,14 @@ class PushController
    */
   void run()
   {
-    if (!running_.load())
+    if (!running_.exchange(true))
     {
-      running_ = true;
       while (active_.load())
       {
         tick();
```

Comment:
Unreated to this PR, but related to concept in general: if you have some way of knowing whether you need to execute the tick or not (i.e. if nothing happened, no new data - then you do not need to wake), you could use a binary semaphore and event-driven notification to tell if you actually need to run the tick. If you skip ticks, you can still tell after how many you skipped because there was no job to do. It'd be best for performance. Note that `sleep_for` may also have spurious wake-ups, and may also be unsteady / sensitive to NTP time clock adjustments in real production environment.

Instead of active poll and tick, you may consider binary semaphore instead:

```cpp
    // Event (binary semaphore)
    class Event
    {
    protected:
        bool m_bFlag;
        mutable std::mutex m_mutex;
        mutable std::condition_variable m_condition;

    public:

        inline Event() : m_bFlag(false) {}

        ~Event()
        {
            Reset();
        }

        bool wait(unsigned millis = UINT_MAX) const
        {
            if (millis == UINT_MAX)
            {
                std::unique_lock< std::mutex > lock(m_mutex);
                m_condition.wait(lock, [&]()->bool {return m_bFlag; });
                return true;
            }

            auto crRelTime = std::chrono::milliseconds(millis);
            std::unique_lock<std::mutex> ulock(m_mutex);
            if (!m_condition.wait_for(ulock, crRelTime, [&]()->bool {return m_bFlag; }))
                return false;
            return true;
        }

        inline bool post()
        {
            bool bWasSignalled;
            m_mutex.lock();
            bWasSignalled = m_bFlag;
            m_bFlag = true;
            m_mutex.unlock();
            m_condition.notify_all();
            return bWasSignalled == false;
        }

        inline bool Reset()
        {
            bool bWasSignalled;
            m_mutex.lock();
            bWasSignalled = m_bFlag;
            m_bFlag = false;
            m_mutex.unlock();
            return bWasSignalled;
        }

        inline bool IsSet() const { return m_bFlag; }
    };
```

Reply:
Thanks max, I'll look into this.  If I can incorporate any of this into the controller I'll redesign it in a future PR.